### ISSUE TITLE
is batch norm calculate formula wrong? 

### ISSUE THREAD

**Issue by dfhe2004** [BVLC/caffe#3912]
According to batch_norm_layer.cpp, the BP calculate formula is :
``
dE(Y)/dX = ( dE/dY-mean(dE/dY)- mean(dE/dY \cdot Y) \cdot Y)
               ./ sqrt(var(X) + eps )
``
but I think it should be:
``
tmp = (dE/dY - mean(dE/dY \cdot Y) \cdot Y) ./ sqrt(var(X) + eps )
``
``
dE(Y)/dX = tmp - mean(tmp)
``

 I also compared the codes with mxnet implement that follows the second one.

**Comment by dfhe2004** [comment]
Sorry, I made a mistake, The above tow formulas are equivalent forms.
